scripts/bert/bert_qa_dataset.py

```diff
@@ -0,0 +1,516 @@
+                 max_query_length=64,
+                 is_training=True):
+        self.tokenizer = tokenizer
+        self.max_seq_length = max_seq_length
+        self.max_query_length = max_query_length
+        self.doc_stride = doc_stride
+        self.is_training = is_training
+
+    def _is_whitespace(self, c):
+        if c == ' ' or c == '\t' or c == '\r' or c == '\n' or ord(
+                c) == 0x202F:
+            return True
+        return False
+
+    def _toSquadExample(self, record):
+        example_id = record[0]
+        qas_id = record[1]
+        question_text = record[2]
+        paragraph_text = record[3]
+        orig_answer_text = record[4][0]
+        answer_offset = record[5][0]
+        is_impossible = record[6] if len(record) == 7 else False
+
+        doc_tokens = []
+
+        char_to_word_offset = []
+        prev_is_whitespace = True
+        for c in paragraph_text:
+            if self._is_whitespace(c):
+                prev_is_whitespace = True
+            else:
+                if prev_is_whitespace:
+                    doc_tokens.append(c)
+                else:
+                    doc_tokens[-1] += c
+                prev_is_whitespace = False
+            char_to_word_offset.append(len(doc_tokens) - 1)
+
+        start_position = None
+        end_position = None
+
+        if self.is_training:
+            if not is_impossible:
+                answer_length = len(orig_answer_text)
+                start_position = char_to_word_offset[answer_offset]
+                end_position = char_to_word_offset[
+                    answer_offset + answer_length - 1]
+                # Only add answers where the text can be exactly recovered from the
+                # document. If this CAN'T happen it's likely due to weird Unicode
+                # stuff so we will just skip the example.
+                #
+                # Note that this means for training mode, every example is NOT
+                # guaranteed to be preserved.
+                actual_text = ' '.join(
+                    doc_tokens[start_position:(end_position + 1)])
+                cleaned_answer_text = ' '.join(
+                    whitespace_tokenize(orig_answer_text))
+                if actual_text.find(cleaned_answer_text) == -1:
+                    print('Could not find answer: %s vs. %s' %
+                          (actual_text, cleaned_answer_text))
+                    return None
+            else:
+                start_position = -1
+                end_position = -1
+                orig_answer_text = ''
+
+        example = SquadExample(
+            qas_id=qas_id,
+            question_text=question_text,
+            doc_tokens=doc_tokens,
+            example_id=example_id,
+            orig_answer_text=orig_answer_text,
+            start_position=start_position,
+            end_position=end_position,
+            is_impossible=is_impossible)
+        return example
+
+    def _transform(self, *record):
```

Comment:
should we change this as the default `__call__` function? since when calling `SQuADTransform `, it only uses this function. we thus could directly return,

Reply:
The result of the _transform function is used when calculating the answer.